src/sources/file.rs

```diff
@@ -561,16 +568,26 @@ pub fn file_source(
 
         let mut encoding_decoder = encoding_charset.map(Decoder::new);
 
+        let mut file_id_to_metrics_mapping =
+            HashMap::<FileFingerprint, FileBytesReceivedHandle>::new();
+
         // sizing here is just a guess
         let (tx, rx) = futures::channel::mpsc::channel::<Vec<Line>>(2);
         let rx = rx
             .map(futures::stream::iter)
             .flatten()
             .map(move |mut line| {
-                emit!(FileBytesReceived {
-                    byte_size: line.text.len(),
-                    file: &line.filename,
-                });
+                let file_bytes_received_entry = file_id_to_metrics_mapping.entry(line.file_id);
+
+                let file_bytes_received_handle = match file_bytes_received_entry {
+                    Entry::Occupied(o) => o.into_mut(),
+                    Entry::Vacant(v) => v.insert(register!(FileBytesReceived {
+                        file: SharedString::from(line.filename.clone())
+                    })),
+                };
+
+                file_bytes_received_handle.emit(ByteSize(line.text.len()));
```

Comment:
Initially, I've implemented it exactly this way. But for some unknown reason seems like `.or_insert_with` was evaluated in each call, regardless the `.entry` call. Why I noticed it because I got a performance regression. After that I reimplemented it as it's done right now - and it works fine.

I could recheck it once again if you wish. The only caveat here is the Release build time - it takes 12 min on my local machine.

Reply:
Hmm, I don't see why, so I'm curious what you find here.

Note that you can dramatically reduce build times with `--no-default-features --features sources-file`, or run `cargo vdev run CONFIG` which will autodetect the needed flags for a given config.

Reply:
@bruceg you are right: I've applied your suggestions and it works like it should, without degradation. I guess I just did something wrong in my previous version. So... it's fixed now and works as fast as it should.

Thanks!